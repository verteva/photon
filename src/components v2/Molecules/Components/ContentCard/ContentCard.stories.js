import PContentCard from './ContentCard.vue';

export default {
  title: 'V2/Molecules/Components/ContentCard',
  component: PContentCard,
  argTypes: {
    // Configurable component options in SB UI...
  },
};

const Template = (args, { argTypes }) => ({
  components: { PContentCard },
  props: Object.keys(argTypes),
  template: `<PContentCard v-bind='$props'
  >
    <div class="ph-flex ph-items-center">
      <div class="ph-font-bold ph-mt-1 ph-leading-5 ph-mr-auto">
        I want a better <br />variable rate
      </div>
    </div>
  </PContentCard>`,
});

export const ContentCard = Template.bind({});
ContentCard.args = {
  // Props to be passed....
};
