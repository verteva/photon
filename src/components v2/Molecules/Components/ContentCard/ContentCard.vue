<template>
  <p2-card
    ref="contentCard"
    class="content-card"
    :size="size"
    :class="[
      {
        disabled,
        active,
        bordered,
        interactive: isInteractive && !disabled,
      },
      type,
    ]"
    :tabindex="tabindex || isInteractive ? 0 : ''"
    @keyup.native.enter="isInteractive && $emit('click')"
    @click.native="isInteractive && $emit('click')"
  >
    <div class="content-card-backdrop"></div>
    <div class="content-card-content">
      <slot />
    </div>
  </p2-card>
</template>

<script lang="ts">
export const props = {
  active: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  bordered: {
    type: Boolean,
    default: false,
  },
  isInteractive: {
    type: Boolean,
    default: false,
  },
  size: {
    type: String,
    default: 'sm',
  },
  tabindex: {
    type: String,
    default: '',
  },
  type: {
    type: String,
    default: '',
  },
};
export default {};
</script>

<script setup lang="ts">
import P2Card from '@/components v2/Atoms/Components/Card';

defineProps(props);
</script>

<style lang="scss" scoped>
.content-card {
  background: var(--sd-card-light-background-color);
  border-radius: var(--sd-card-default-border-radius);
  transition: all ease-in-out 300ms;
  position: relative;
  border: 1px solid transparent;

  .content-card-content {
    position: relative;
    z-index: 1;
    transition: color ease-in-out 300ms;
    width: 100%;
  }

  .content-card-backdrop {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
    transition: opacity 300ms;
    background: var(--sd-theme-primary-subtle);
    border-radius: var(--sd-card-default-border-radius);
    overflow: hidden;

    &:after {
      background: var(--sd-theme-primary-default);
      animation-fill-mode: both;
      transition: transform 0.35s cubic-bezier(0.3, 0.28, 0.26, 0.81);
      animation-direction: normal, reverse;
      transform-origin: 50% 100%;
      content: '';
      width: 100%;
      height: 4px;
      bottom: 0;
      left: 0;
      position: absolute;
      transform: scaleX(0) translateY(4px);
      z-index: 1;
      border-bottom-left-radius: var(--sd-card-default-border-radius);
      border-bottom-right-radius: var(--sd-card-default-border-radius);
    }
  }

  &.bordered {
    border-color: var(--sd-theme-input-border-color);
  }

  &.interactive {
    cursor: pointer;

    &:hover {
      .content-card-backdrop {
        opacity: 1;
      }
    }

    &:focus,
    &:focus-visible,
    &:focus-within {
      outline: none;
      color: var(--sd-input-text-focus-text-color);
      background: var(--sd-input-text-focus-background-color);
      box-shadow: var(--sd-input-text-focus-box-shadow);
      border-color: var(--sd-input-text-focus-border-color);
    }

    &.bordered {
      &:hover {
        border: 1px solid var(--sd-theme-primary-default);
      }

      .content-card-backdrop {
        background: var(--sd-theme-primary-subtle);
      }
    }

    &.active {
      border-color: var(--sd-theme-primary-default);

      .content-card-backdrop {
        background: var(--sd-theme-primary-subtle);
        opacity: 1;
        &:after {
          transform: scaleX(1) translateY(0);
        }
      }

      .content-card-content {
        color: var(--sd-theme-primary-default);
      }
    }
  }

  &.disabled {
    cursor: not-allowed;
  }
}

@keyframes grow-out {
  0% {
    transform: scaleX(0);
  }
  100% {
    transform: scaleX(1);
  }
}
</style>
