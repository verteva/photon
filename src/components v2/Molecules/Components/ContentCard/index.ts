export { default, props } from './ContentCard.vue';
