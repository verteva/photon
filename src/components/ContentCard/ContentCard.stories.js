import PCard from './ContentCard.vue';
import PIcon from '../Icon'; 

export default {
  title: 'Components/ContentCard',
  component: PCard,
  argTypes: {
    // Configurable component options in SB UI...
  },
};

const Template = (args, { argTypes }) => ({
  components: { PContentCard, PIcon },
  props: Object.keys(argTypes),
  template: `<PContentCard v-bind='$props' 
  >      
    <div class="ph-flex ph-items-center">
      <div class="ph-font-bold ph-mt-1 ph-leading-5 ph-mr-auto ">
        I want a better <br />variable rate
      </div>
      <p-icon
        name="Percentage"
        type="xl"
        class="ph-text-brand2"
      />
    </div>
  </PContentCard>`,
});

export const ContentCard = Template.bind({});
ContentCard.args = {
  // Props to be passed....
};
