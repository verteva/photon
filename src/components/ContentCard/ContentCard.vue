<template>
  <div
    ref="contentCard"
    v-bind="$attrs"
    class="content-card ph-bg-gradient-light-grey2 ph-rounded-lg ph-transition-opacity ph-duration-300 ph-relative ph-border"
    :class="(defaultParentClass, conditionParentClass)"
    v-on="$listeners"
    @mouseover="mouseOver"
    @mouseout="mouseOver"
  >
    <div :class="(conditionChildClass, defaultChildClass)" />
    <div
      class="ph-relative ph-z-1 ph-transition-color ph-w-full"
      :class="[active && 'ph-text-brand2', padding]"
    >
      <slot />
    </div>
  </div>
</template>
<script lang="ts">
export default {
  name: 'ContentCard',
  props: {
    active: {
      type: Boolean,
      default: false
    },
    flex: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    isInteractive: {
      type: Boolean,
      default: false
    },
    padding: {
      type: String,
      default: 'ph-p-7'
    },
    allowOverflow: {
      type: Boolean,
      default: false
    },
    breakpoint: {
      type: Boolean,
      default: false
    }
  },
  data(): any {
    return {
      hover: false
    };
  },
  computed: {
    interactive(): any {
      if ((this as any).isInteractive) return true;
      else
        return (
          (this as any).$listeners &&
          (this as any).$listeners.click &&
          !(this as any).disabled
        );
    },
    conditionParentClass(): any {
      return [
        (this as any).active ? 'ph-border-brand2 bar-bottom-brand' : 'ph-border-grey9',
        (this as any).interactive && !(this as any).disabled && 'ph-cursor-pointer',
        !(this as any).flex && 'ph-h-height',
        (this as any).allowOverflow ? 'ph-overflow-visible' : 'ph-overflow-hidden'
      ];
    },
    defaultParentClass(): any {
      return [
        'content-card',
        'ph-bg-gradient-light-grey2',
        'ph-rounded-lg',
        'ph-transition-opacity',
        'ph-duration-300',
        'ph-relative ph-border'
      ];
    },
    conditionChildClass(): any {
      return [
        (this as any).interactive &&
          !(this as any).disabled &&
          (this as any).hover &&
          (this as any).breakpoint &&
          'ph-opacity-5',
        (this as any).interactive && !(this as any).disabled && (this as any).active && 'ph-opacity-10'
      ];
    },
    defaultChildClass(): string[] {
      return [
        'ph-absolute',
        'ph-top-0',
        'ph-left-0',
        'ph-right-0',
        'ph-bottom-0',
        'ph-opacity-0',
        'ph-transition-opacity',
        'ph-duration-300',
        'ph-bg-brand2'
      ];
    }
  },
  methods: {
    mouseOver(): void {
      (this as any).hover = !(this as any).hover;
    }
  }
};
</script>

<style lang="postcss" scoped>
.content-card.bar-bottom-brand:after {
  @apply ph-bg-brand2;
  @apply ph-absolute;
  animation-fill-mode: both;
  transform-origin: 50% 100%;
  content: '';
  width: 100%;
  height: 6px;
  bottom: 0;
  left: 0;
  @apply ph-animate-growOut;
}
</style>
