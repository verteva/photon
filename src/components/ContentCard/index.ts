import PContentCard from './ContentCard.vue';

export default PContentCard;
