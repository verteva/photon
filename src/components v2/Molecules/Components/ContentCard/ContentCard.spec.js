import { render } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import ContentCard from './ContentCard';

const props = {
  active: true,
  disabled: true,
  isInteractive: true,
  allowOverflow: true,
  type: 'bordered',
};

describe('ContentCard.vue', () => {
  it('renders a component', () => {
    const { container } = render(ContentCard);
    expect(container).toMatchSnapshot();
  });

  Object.entries(props).forEach(([key, value]) => {
    it(`renders a component with props: ${key} with value ${value}`, () => {
      const { container } = render(ContentCard, {
        props: {
          [key]: value,
        },
      });
      expect(container).toMatchSnapshot();
    });
  });

  it('slots: renders default slot', () => {
    const slotText = 'I am text';
    const { getByText } = render(ContentCard, {
      slots: {
        default: slotText,
      },
    });

    const content = getByText(slotText);

    expect(content).toBeTruthy();
  });

  it('emits: click when pressed', async () => {
    const slotText = 'I am text';
    const { getByText, emitted } = render(ContentCard, {
      slots: {
        default: slotText,
      },
    });

    const content = getByText(slotText);

    await userEvent.click(content);

    expect(emitted().click).toBeTruthy();
  });
});
