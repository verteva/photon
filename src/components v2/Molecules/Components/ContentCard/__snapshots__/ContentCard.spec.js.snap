// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentCard.vue renders a component 1`] = `
<div>
  <div
    class="card content-card rounded no-padding"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;

exports[`ContentCard.vue renders a component with props: active with value true 1`] = `
<div>
  <div
    class="card content-card rounded no-padding active"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;

exports[`ContentCard.vue renders a component with props: allowOverflow with value true 1`] = `
<div>
  <div
    class="card content-card rounded no-padding allow-overflow"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;

exports[`ContentCard.vue renders a component with props: disabled with value true 1`] = `
<div>
  <div
    class="card content-card rounded no-padding disabled"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;

exports[`ContentCard.vue renders a component with props: isInteractive with value true 1`] = `
<div>
  <div
    class="card content-card rounded no-padding interactive"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;

exports[`ContentCard.vue renders a component with props: type with value bordered 1`] = `
<div>
  <div
    class="card content-card rounded no-padding bordered"
  >
    <div
      class="content-card-backdrop"
    />
     
    <div
      class="content-card-content"
    />
  </div>
</div>
`;
